README.md
# MssqlEcto

[![Build Status](https://travis-ci.org/findmypast-oss/mssql_ecto.svg?branch=master)](https://travis-ci.org/findmypast-oss/mssql_ecto)
[![Coverage Status](https://coveralls.io/repos/github/findmypast-oss/mssql_ecto/badge.svg)](https://coveralls.io/github/findmypast-oss/mssql_ecto)
[![Inline docs](http://inch-ci.org/github/findmypast-oss/mssql_ecto.svg?branch=master)](http://inch-ci.org/github/findmypast-oss/mssql_ecto)
[![Ebert](https://ebertapp.io/github/findmypast-oss/mssql_ecto.svg)](https://ebertapp.io/github/findmypast-oss/mssql_ecto)
[![Hex.pm](https://img.shields.io/hexpm/v/mssql_ecto.svg)](https://hex.pm/packages/mssql_ecto)
[![LICENSE](https://img.shields.io/hexpm/l/mssql_ecto.svg)](https://github.com/findmypast-oss/mssql_ecto/blob/master/LICENSE)

[Ecto](https://github.com/elixir-ecto/ecto) Adapter for [Mssqlex](https://github.com/findmypast-oss/mssqlex)

## Installation

### Erlang ODBC Application

MssqlEcto requires the [Erlang ODBC application](http://erlang.org/doc/man/odbc.html) to be installed. This might require the installation of an additional package depending on how you have installed Elixir/Erlang (e.g. on Ubuntu `sudo apt-get install erlang-odbc`).

### Microsoft's ODBC Driver

MssqlEcto depends on Microsoft's ODBC Driver for SQL Server. You can find installation instructions for [Linux](https://docs.microsoft.com/en-us/sql/connect/odbc/linux/installing-the-microsoft-odbc-driver-for-sql-server-on-linux) or [other platforms](https://docs.microsoft.com/en-us/sql/connect/odbc/microsoft-odbc-driver-for-sql-server) on the official site.

### Hex

#### With [Application Inference](https://elixir-lang.org/blog/2017/01/05/elixir-v1-4-0-released/#application-inference)

If you are using [application inference](https://elixir-lang.org/blog/2017/01/05/elixir-v1-4-0-released/#application-inference), i.e. `application` in your `mix.exs` looks something like this:

```elixir
def application do
  [extra_applications: [:logger]]
end
```

Note, the lack of `:applications` key. Then, you just need to add the following dependencies:

```elixir
def deps do
  [{:mssql_ecto, "~> 1.1.0"},
   {:mssqlex, "~> 1.0.0"}]
end
```

#### Without [Application Inference](https://elixir-lang.org/blog/2017/01/05/elixir-v1-4-0-released/#application-inference)

If you are explicitly calling out all running applications under `application` in your `mix.exs`, i.e. it looks something like this:

```elixir
def application do
  [applications: [:logger, :plug, :postgrex]]
end
```

Then, you need to add `mssql_ecto` and `mssqlex` to both your `deps` and list of running applications:

```elixir
def application do
  [applications: [:logger, :plug, :mssqlex, :mssql_ecto]]
end

def deps do
  [{:mssql_ecto, "~> 1.1.0"},
   {:mssqlex, "~> 1.0.0"}]
end
```

## Configuration

Example configuration:

```elixir
config :my_app, MyApp.Repo,
  adapter: MssqlEcto,
  database: "sql_server_db",
  username: "bob",
  password: "mySecurePa$$word",
  hostname: "localhost",
  instance_name: "MSSQLSERVER",
  port: "1433"
```

## Example Project

An example project using mssql_ecto with Docker has kindly been created by [Chase Pursłey](https://github.com/cpursley). It can be viewed [here](https://github.com/cpursley/mssql_ecto_friends).

## Type Mappings

|    Ecto Type    |   SQL Server Type    |                                                                       Caveats                                                                        |
| :-------------: | :------------------: | :--------------------------------------------------------------------------------------------------------------------------------------------------: |
|       :id       |         int          |                                                                                                                                                      |
|     :serial     |  int identity(1, 1)  |                                                                                                                                                      |
|   :bigserial    | bigint identity(1,1) | When a query is returning this value with the `returning` syntax and no schema is used, it will be returned as a string rather than an integer value |
|   :binary_id    |       char(36)       |                                                                                                                                                      |
|      :uuid      |       char(36)       |                                                                                                                                                      |
|     :string     |       nvarchar       |                                                                                                                                                      |
|     :binary     |    nvarchar(4000)    |                                                         Limited size, not fully implemented                                                          |
|    :integer     |         int          |                                                                                                                                                      |
|    :boolean     |         bit          |                                                                                                                                                      |
| {:array, type}  |     list of type     |                                                                    Not Supported                                                                     |
|      :map       |    nvarchar(4000)    |                                                                    Not Supported                                                                     |
|   {:map, \_}    |    nvarchar(4000)    |                                                                    Not Supported                                                                     |
|      :date      |         date         |                                                                                                                                                      |
|      :time      |         time         |                                                               Can write but can't read                                                               |
|  :utc_datetime  |      datetime2       |                                                                                                                                                      |
| :naive_datetime |      datetime2       |                                                                                                                                                      |
|     :float      |        float         |                                                                                                                                                      |
|    :decimal     |       decimal        |                                                                                                                                                      |

## Features not yet implemented

* Table comments
* Column comments
* On conflict
* Upserts

## Contributing

### Code of Conduct

This project had a [Code of Conduct](https://github.com/findmypast-oss/mssql_ecto/blob/master/CODE_OF_CONDUCT.md) if you wish to contribute to this project, please abide by its rules.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2018-02-12

### Changed

* Updating to `ecto` verion 2.2

## [1.0.0] - 2018-01-31

### Changed

* Upping version of `mssqlex` to 1.0.0, thus allowing the following change:
  * Changed the default version of the ODBC Driver to 17. This is to reflect what version is installed when running apt-get install msodbcsql on Debian Jessie. It may cause breaking changes for some users who rely on the default being 13.

## [0.3.0] - 2017-07-21

### Added

* Upping version of `mssqlex` to 0.8.0, thus allowing the named instance option.

## [0.2.0] - 2017-07-06

### Added

* Upping version of `mssqlex` to 0.7.0 thus allowing custom ports.


lib/mssql_ecto/helpers.ex
defmodule MssqlEcto.Helpers do
  alias MssqlEcto.QueryString

  def get_source(query, sources, ix, source) do
    {expr, name, _schema} = elem(sources, ix)
    {expr || QueryString.paren_expr(source, sources, query), name}
  end

  def quote_qualified_name(name, sources, ix) do
    {_, source, _} = elem(sources, ix)
    [source, ?. | quote_name(name)]
  end

  def quote_name(name, quoter \\ ?")
  def quote_name(nil, _), do: []

  def quote_name(names, quoter) when is_list(names) do
    names
    |> Enum.filter(&(not is_nil(&1)))
    |> intersperse_map(?., &quote_name(&1, nil))
    |> wrap_in(quoter)
  end

  def quote_name(name, quoter) when is_atom(name) do
    quote_name(Atom.to_string(name), quoter)
  end

  def quote_name(name, quoter) do
    if String.contains?(name, "\"") do
      error!(nil, "bad name #{inspect(name)}")
    end

    wrap_in(name, quoter)
  end

  def wrap_in(value, nil), do: value

  def wrap_in(value, {left_wrapper, right_wrapper}) do
    [left_wrapper, value, right_wrapper]
  end

  def wrap_in(value, wrapper) do
    [wrapper, value, wrapper]
  end

  def quote_table(prefix, name)
  def quote_table(nil, name), do: quote_name(name)

  def quote_table(prefix, name),
    do: intersperse_map([prefix, name], ?., &quote_name/1)

  def single_quote(value), do: value |> escape_string |> wrap_in(?')

  def intersperse_map(list, separator, mapper, acc \\ [])
  def intersperse_map([], _separator, _mapper, acc), do: acc

  def intersperse_map([elem], _separator, mapper, acc),
    do: [acc | mapper.(elem)]

  def intersperse_map([elem | rest], separator, mapper, acc),
    do:
      intersperse_map(rest, separator, mapper, [acc, mapper.(elem), separator])

  def intersperse_reduce(list, separator, user_acc, reducer, acc \\ [])

  def intersperse_reduce([], _separator, user_acc, _reducer, acc),
    do: {acc, user_acc}

  def intersperse_reduce([elem], _separator, user_acc, reducer, acc) do
    {elem, user_acc} = reducer.(elem, user_acc)
    {[acc | elem], user_acc}
  end

  def intersperse_reduce([elem | rest], separator, user_acc, reducer, acc) do
    {elem, user_acc} = reducer.(elem, user_acc)

    intersperse_reduce(rest, separator, user_acc, reducer, [
      acc,
      elem,
      separator
    ])
  end

  def if_do(condition, value) do
    if condition, do: value, else: []
  end

  def escape_string(value) when is_binary(value) do
    :binary.replace(value, "'", "''", [:global])
  end

  def ecto_to_db({:array, t}), do: [ecto_to_db(t), ?[, ?]]
  def ecto_to_db(:id), do: "int"
  def ecto_to_db(:serial), do: "int identity(1,1)"
  def ecto_to_db(:bigserial), do: "bigint identity(1,1)"
  def ecto_to_db(:binary_id), do: "char(36)"
  def ecto_to_db(:uuid), do: "char(36)"
  def ecto_to_db(:string), do: "nvarchar"
  def ecto_to_db(:binary), do: "nvarchar(4000)"
  def ecto_to_db(:integer), do: "int"
  def ecto_to_db(:boolean), do: "bit"
  def ecto_to_db(:map), do: "nvarchar(4000)"
  def ecto_to_db({:map, _}), do: "nvarchar(4000)"
  def ecto_to_db(:utc_datetime), do: "datetime2"
  def ecto_to_db(:naive_datetime), do: "datetime2"
  def ecto_to_db(:timestamp), do: "datetime2"
  def ecto_to_db(other), do: Atom.to_string(other)

  def error!(nil, message) do
    raise ArgumentError, message
  end

  def error!(query, message) do
    raise Ecto.QueryError, query: query, message: message
  end
end
